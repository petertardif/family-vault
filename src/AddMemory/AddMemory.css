.add-memory-container {
    width: 50%;
    margin: 200px auto 10px auto;
    opacity: 0.8;
    border-radius: 13px;
    padding: 1em;
    min-width: 300px;
    border: 1px solid grey;
}

.add-memory-form input,
.add-memory-form textarea,
.add-memory-form select {
  font-size: 14px;
  padding: 8px;
  width: 75%;
  /* min-width: 200px; */
}

input[type="date"]::-webkit-calendar-picker-indicator {
    color: #2c3e50;
}

/* A few custom styles for date inputs */
input[type="date"] {
    appearance: none;
    -webkit-appearance: none;
    color: #444;
    font-size: 18px;
    font-family: 'Open Sans', sans-serif;
    border: 1px solid lightgrey;
    padding:5px;
    visibility: visible !important;
}

mobile styles
@media only screen and (min-width: 350px) {
  .add-memory-container {
    width: 50%;
    margin: 200px auto 10px auto;
    opacity: 0.8;
    border-radius: 13px;
    padding: 1em;
    /* min-width: 300px; */
    border: 1px solid grey;

}

  .add-memory-form input,
  .add-memory-form textarea,
  .add-memory-form select {
    font-size: 14px;
    padding: 8px;
    width: 350px;
  }
}
