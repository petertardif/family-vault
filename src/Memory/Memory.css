.Memory {
    margin: 5px auto;
    opacity: 0.8;
    border-radius: 13px;
    padding: 1em;
    display: flex;
    justify-content: center;
    border: 1px gray solid;
}

@media only screen and (min-width: 320px) {
    .Memory {
        width: 75%;
        min-width: 300px;
        padding: 10px;
    }
}

@media only screen and (min-width: 540px) {
    .Memory {
        width: 65%;
    }
}

@media only screen and (min-width: 800px) {
    .Memory {
        width: 55%;
    }
}