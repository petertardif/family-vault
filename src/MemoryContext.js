import React from 'react';

export const MemoryContext = React.createContext({
  on: '', 
  deleteOn: '',
  toggle: () => {},
  deleteToggle: () => {},
  memories: [],
  familyMembers: [],
  addMemory: () => {},
  deleteMemory: () => {},
  addFamilyMember: () => {},
  deleteFamilyMember: () => {},
  updateMemoryList: () => {},
});