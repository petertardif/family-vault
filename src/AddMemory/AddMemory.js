import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import MemoryForm from '../MemoryForm/MemoryForm';
import { MemoryContext } from '../MemoryContext';
import './AddMemory.css';
import { API_BASE_URL, AWSAccessKeyId, AWSSecretKey } from '../config';
import S3FileUpload from 'react-s3';

const config = {
  bucketName: 'familyvaultapi',
  region: 'us-east-2',
  accessKeyId: AWSAccessKeyId,
  secretAccessKey: AWSSecretKey,
}

class AddMemory extends Component {
  static defaultProps = {
    familyMembers: [],
    history: {
      push: () => { }
    }
  }

static contextType = MemoryContext;

constructor(props) {
  super(props);
  this.state = {
    memoryTitle: '',
    memoryDescription: '',
    memoryFamilyMember: '',
    memoryMedia: '',
    memoryDate: '',
    memoryTitleValid: false,
    memoryDescriptionValid: false,
    memoryFamilyMemberValid: false,
    memoryDateValid: false,
    formValid: false,
    validationMessages: {
      memoryTitle: '',
      memoryDescription: '',
      memoryFamilyMember: '',
      memoryDate: '',
    }
  }
}

  updateMemoryTitle(memoryTitle) {
    this.setState({memoryTitle}, () => {this.validateMemoryTitle(memoryTitle)});
  }

  updateMemoryDescription(memoryDescription) {
    this.setState({memoryDescription}, () => {this.validateMemoryDescription(memoryDescription)});
  }

  updateFamilyMember(memoryFamilyMember) {
    this.setState({memoryFamilyMember}, () => {this.validateFamilyMember(memoryFamilyMember)});
  }

  updateMemoryMedia = (memoryMedia) => {
    console.log(memoryMedia.target.files[0]);
    S3FileUpload
      .uploadFile(memoryMedia.target.files[0], config)
      .then( (data) => {
        this.setState({memoryMedia: data.location})
      })
      .catch(err => console.error(err))
  }

  updateMemoryDate(memoryDate) {
    this.setState({memoryDate}, () => {this.validateMemoryDate(memoryDate)});
  }

  validateMemoryTitle(fieldValue) {
    const fieldErrors = {...this.state.validationMessages};
    let hasError = false;

    fieldValue = fieldValue.trim();
    if(fieldValue.length === 0) {
      fieldErrors.memoryTitle = 'Please type a memory title';
      hasError = true;
    }

    this.setState({
      validationMessages: fieldErrors,
      memoryTitleValid: !hasError
    }, this.formValid );
  }

  validateMemoryDescription(fieldValue) {
    const fieldErrors = {...this.state.validationMessages};
    let hasError = false;

    fieldValue = fieldValue.trim();
    if(fieldValue.length === 0) {
      fieldErrors.memoryDescription = 'Please type a description of the memory ';
      hasError = true;
    }

    this.setState({
      validationMessages: fieldErrors,
      memoryDescriptionValid: !hasError
    }, this.formValid );
  }

  validateFamilyMember(fieldValue) {
    const fieldErrors = {...this.state.validationMessages};
    let hasError = false;

    if(fieldValue === "empty") {
      fieldErrors.memoryFamilyMember = 'Please select a Family Member';
      hasError = true;
    }

    this.setState({
      validationMessages: fieldErrors,
      memoryFamilyMemberValid: !hasError
    }, this.formValid );
  }

  validateMemoryDate(fieldValue) {
    const fieldErrors = {...this.state.validationMessages};
    let hasError = false;

    fieldValue = fieldValue.trim();
    if(fieldValue.length === 0) {
      fieldErrors.memoryDate = 'Please enter a valid date.';
      hasError = true;
    }

    this.setState({
      validationMessages: fieldErrors,
      memoryDateValid: !hasError
    }, this.formValid );
  }

  formValid() {
    this.setState({
      formValid: this.state.memoryTitleValid && this.state.memoryDescriptionValid && this.state.memoryFamilyMemberValid &&
      this.state.memoryDateValid
    });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const newMemory = {
      memory_title: e.target['memory-title'].value,
      memory_desc: e.target['memory-description'].value,
      familymember_id: e.target['family-member-id'].value,
      media_url: this.state.memoryMedia,
      memory_date: e.target['memory-date'].value,
      date_updated: new Date().toDateString(),
    }
    console.log(newMemory);

    fetch(`${API_BASE_URL}/memories`, {
      method: 'POST',
      headers: {
        'content-type': 'application/json'
      },
      body: JSON.stringify(newMemory),
    })
      .then(res => {
        if (!res.ok)
          return res.json().then(e => Promise.reject(e))
        return res.json()
      })
      .then(memory => {
        this.context.addMemory(memory)
        this.props.history.push(`/memorylist`)
      })
      .catch(error => {
        console.error({ error })
      })
  }

  handleClickGoBack = () => {
    this.props.history.push('/userlanding');
  } 
  
  render() {
    const { familyMembers = [] } = this.context;
    // variable below limits the user from selecting a date in the future.
    const today = new Date().toISOString().substr(0, 10);
    
    return (
      <div>
        <section className='add-memory-container'>
          <h1>New Memory</h1>
          <MemoryForm onSubmit={e => this.handleSubmit(e)}>
            <div className='add-memory-form '>
              <label htmlFor='memory-title-input'>
                Name it
              </label>
              <input type='text' id='memory-title-input' name='memory-title' onChange={e => this.updateMemoryTitle(e.target.value)} />
            </div>
            <div className='add-memory-form '>
              <label htmlFor='memory-description-input'>
                Describe it
              </label>
              <textarea id='memory-description-input' name='memory-description' onChange={e => this.updateMemoryDescription(e.target.value)} />
            </div>
            <div className='add-memory-form '>
              <label htmlFor='family-member-select'>
                Family Member
              </label>
              <select id='family-member-select' name='family-member-id' onChange={e => this.updateFamilyMember(e.target.value)}>
                <option value="empty">...</option>
                {familyMembers.map(fm =>
                  <option key={fm.id} value={fm.id}>
                    {fm.first_name} {fm.last_name}
                  </option>
                )}
              </select>
            </div>
            <div className='add-memory-form '>
              <label htmlFor='memory-media-input'>
                Add pic/video
              </label>
              <input type='file' id='memory-media-input' name='memory-media' onChange={this.updateMemoryMedia} />
            </div>
            <div className='add-memory-form '>
              <label htmlFor='memory-date-input'>
                Date
              </label>
              <input type='date' id='memory-date-input' max={today} name='memory-date' onChange={e => this.updateMemoryDate(e.target.value)} />
            </div>
            <div className='buttons add-memory-buttons'>
              <button type='submit' disabled={!this.state.formValid} className='Button blue'>
                Add memory
              </button>
              <button onClick={this.handleClickGoBack} className='Button blue'>Back</button>
            </div>
          </MemoryForm>
        </section>
      </div>
    )
  }
}

export default withRouter(AddMemory);
