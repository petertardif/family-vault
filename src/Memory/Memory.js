import React, { Component } from 'react';
import { Link, withRouter } from 'react-router-dom';
import Toggle from '../Toggle/Toggle';
import { API_BASE_URL } from '../config';
import './Memory.css';
import { MemoryContext } from '../MemoryContext';

class Memory extends Component {
  // static defaultProps = {
  //   onDeleteMemory: () => {},
  // }
  static contextType = MemoryContext;

  handleDeleteRequest = (event) => {
    event.preventDefault()
    const memoryId = this.props.id
    debugger
    fetch(`${API_BASE_URL}/memories/${memoryId}`, {
      method: 'DELETE',
      headers: {
        'content-type': 'application/json'
      },
    })
      .then(res => {
        if (!res.ok)
          return res.json().then(e => Promise.reject(e))
        return res.json()
      })
      .then(() => {
        this.context.deleteMemory(memoryId)
        this.props.history.push(`/memorylist/${memoryId}`)
      })
      .catch(error => {
        console.error({ error })
      })
      this.props.deleteMemory(memoryId)
      debugger
      this.props.history.push(`/memorylist`)
      // this.props.history.push(`/memorylist/${memoryId}`)
    };
  

  render() {
    return (
      <li key={this.props.id} className='Memory'>
        <div className='Memory_row'>
          <h2 className='Memory_title'>
            <Link to={`/memory/${this.props.id}`}>
              {this.props.memory_title}
              </Link>
          </h2>
          <div className='Memory_dates'>
            <div className='Memory_dates-modified'>
                Modified <span className='Date'>{this.props.memory_date}</span> 
            </div>
          </div>
          <Toggle>
            {({ on, toggle }) => (
              <>
                {on && 
                  <>
                    <img src={this.props.media_url} alt={`supporting visual for ${this.props.title}`}/>
                    <p className='Memory_description'>
                      {this.props.memory_desc}
                    </p>
                    <div className='Memory_buttons'>
                      {/* <button
                        className='Btn_memory_edit'
                        // onClick={() => props.onClickEdit(props.id)}
                      >
                        Edit
                      </button> */}
                      <button
                        type='button'
                        className='Btn_memory_delete'
                        onClick={this.handleDeleteRequest}
                      >
                        Delete
                      </button>
                    </div>
                  </>
                }
                <button onClick={toggle}>Expand / Collapse</button>
              </>
            )}
          </Toggle>
        </div>
      </li>
    )
  }
}

Memory.defaultProps = {
  onClickDelete: () => {},
}

export default withRouter(Memory);