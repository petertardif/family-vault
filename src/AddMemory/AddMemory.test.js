import React from 'react';
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import AddMemory from './AddMemory';
import Memory from '../Memory/Memory';
import { MemoryContext } from '../MemoryContext';

describe(`AddMemory component`, () => {
  it('renders empty given no items', () => {
    const wrapper = shallow(<AddMemory />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('renders memories when supplied with an items prop', () => {
    const testList = [
      { memory_title: 'test item 1',  },
      { memory_title: 'another test item', },
      { memory_title: 'test item number three',  },
    ]
    const wrapper = shallow(<Memory items={testList} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})