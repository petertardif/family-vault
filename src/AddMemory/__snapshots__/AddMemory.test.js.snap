// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMemory component renders empty given no items 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`AddMemory component renders memories when supplied with an items prop 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
